### PR TITLE
Fix deadlock in P2P restarts

### LINKED ISSUE #8088: P2P shuffle restart may deadlock if no running workers exist
An edge case exists where attempting to restart a P2P shuffle may deadlock. 

In this scenario, there are some non-running workers holding inputs to the shuffle, but no more running workers when the shuffle is restarted. This will cause a `processing` `shuffle-transfer` to transition `processing -> released -> waiting -> waiting -> no-worker -> released`. The task will then remain stuck in `released`, causing a deadlock.

There are two possible ways of fixing this: 
1. Carefully design the restarting mechanism to avoid this transition chain.
2. Recommend `waiting` when transitioning a task `no-worker -> released` if it has tasks or clients waiting on its results (similar to the `processing -> released` and `waiting -> released`.

**Reproducer:**

```python
@gen_cluster(client=True, nthreads=[])
async def test_restart_deadlock(c, s):
    async with Worker(s.address) as a:
        async with Nanny(s.address) as b:
            with dask.annotate(workers=[a.address]):
                df = dask.datasets.timeseries(
                    start="2000-01-01",
                    end="2000-03-01",
                    dtypes={"x": float, "y": float},
                    freq="10 s",
                )
            out = dd.shuffle.shuffle(df, "x", shuffle="p2p")
            fut = c.compute(out.x.size)
            await wait_until_worker_has_tasks(
                "shuffle-transfer", b.worker_address, 1, s
            )
            a.status = Status.paused
            while len(s.running) > 1:
                await asyncio.sleep(0.01) 
            await b.process.process.kill()

            while s.running:
                await asyncio.sleep(0.01)


            a.status = Status.running
            await fut
```

### PR DESCRIPTION
**Pull request by hendrikmakait**
Closes #8088

- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

### CODE REVIEW

File: distributed/scheduler.py [hunk 1]
```diff
@@ -2707,14 +2707,10 @@ def transition_no_worker_released(self, key: str, stimulus_id: str) -> RecsMsgs:
...
-        ts.state = "released"
-
-        for dts in ts.dependencies:
-            dts.waiters.discard(ts)
-
-        ts.waiters.clear()
 
-        return {}, {}, {}
+        recommendations: Recs = {}
+        self._propagate_released(ts, recommendations)
```
Comment (hendrikmakait) [inline distributed/scheduler.py:2712]: Writing a dedicated test for this is harder than anticipated. I think we might catch most (all?) non-P2P edge cases in other places. FWIW, removing this from `transition_queued_released` only fails P2P as well.

Comment (fjetter) [reply distributed/scheduler.py:2712]: Well, this kind of test is triggering this transition

```python
@gen_cluster(nthreads=[("", 1)] * 2, client=True)
async def test_no_worker_released(c, s, a, b):
    f1 = c.submit(inc, 1, workers=[a.address], allow_other_workers=True, key='f1')
    f2 = c.submit(inc, f1, resources={"C": 1}, key='f2')

    while not f2.key in s.tasks or s.tasks[f2.key].state != "no-worker":
        await asyncio.sleep(0.01)

    assert f1.key in a.data and not f1.key in b.data
    await a.close()
    ...
```

In this test, `f2` is in `no-worker` and then transitioned to released. The reason why this hasn't caused any issues so far is because in this example, the transition is `no-worker->released->waiting`

The scheduler extension is however expecting two transitions
- <whatever>->released
- released->waiting->processing

i.e. the scheduler extension issues a transition to _released_ and is relying on the recommendation system to trigger appropriate follow ups. All other ordinary transitions rather transition immediately to _waiting_

So, another fix could be to make the shuffle extension smarter and more aware about what the intended target state is. However, I believe that defeats the purpose of the transition engine and this fix is fine. (this ambiguity was/is also a common problem in the worker)

**Review by fjetter — APPROVED** [review]

File: distributed/scheduler.py (hunk 1 above)
Comment (hendrikmakait) [reply distributed/scheduler.py:2712]: > In this test, `f2` is in `no-worker` and then transitioned to released. The reason why this hasn't caused any issues so far is because in this example, the transition is `no-worker->released->waiting`

That's what I meant, sorry. It's hard to write a dedicated test that _fails_ on `main` but isn't using P2P for the reason you described above.



> So, another fix could be to make the shuffle extension smarter and more aware about what the intended target state is. However, I believe that defeats the purpose of the transition engine and this fix is fine. (this ambiguity was/is also a common problem in the worker)

+1, I think it's better to have the reconciliation logic within the state machine instead of forcing all users of the state machine to handle those edge cases.
